### ISSUE TITLE
PrecondionerFactory clashes with the one in dune-istl

### ISSUE THREAD

**Issue by blattms** [OPM/opm-simulators#2266]
For version 2.7 Dune introduced its own Factories for precondtioners and solvers. Unfortunately,
ours are defined in the namespace Dune, too, and the number of template arguments differs:
```
In file included from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/PreconditionerFactory.hpp:26,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/tests/test_preconditionerfactory.cpp:29:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp:45:7: error: conflicting declaration of template ‘template<class Operator, class Comm> struct Dune::PreconditionerFactory’
 class PreconditionerFactory;
       ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/mblatt/src/dune/dune-2.7/dune-istl/dune/istl/solvers.hh:31,
                 from /home/mblatt/src/dune/dune-2.7/dune-istl/dune/istl/schwarz.hh:23,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningBlockPreconditioner.hpp:25,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/PreconditionerFactory.hpp:25,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/tests/test_preconditionerfactory.cpp:29:
/home/mblatt/src/dune/dune-2.7/dune-istl/dune/istl/solverfactory.hh:89:102: note: previous declaration ‘template<class M, class X, class Y> using PreconditionerFactory = Dune::Singleton<Dune::ParameterizedObjectFactory<std::shared_ptr<Dune::Preconditioner<X, Y> >(const M&, const Dune::ParameterTree&)> >’
   using PreconditionerFactory = Singleton<ParameterizedObjectFactory<PreconditionerSignature<M,X,Y>>>;
                                                                                                      ^
In file included from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/PreconditionerFactory.hpp:26,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/tests/test_preconditionerfactory.cpp:29:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp:66:74: error: wrong number of template arguments (2, should be 3)
     using PrecFactory = PreconditionerFactory<OperatorType, Communication>;
                                                                          ^
In file included from /home/mblatt/src/dune/dune-2.7/dune-istl/dune/istl/solvers.hh:31,
                 from /home/mblatt/src/dune/dune-2.7/dune-istl/dune/istl/schwarz.hh:23,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningBlockPreconditioner.hpp:25,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/PreconditionerFactory.hpp:25,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/tests/test_preconditionerfactory.cpp:29:
/home/mblatt/src/dune/dune-2.7/dune-istl/dune/istl/solverfactory.hh:89:102: note: provided for ‘template<class M, class X, class Y> using PreconditionerFactory = Dune::Singleton<Dune::ParameterizedObjectFactory<std::shared_ptr<Dune::Preconditioner<X, Y> >(const M&, const Dune::ParameterTree&)> >’
   using PreconditionerFactory = Singleton<ParameterizedObjectFactory<PreconditionerSignature<M,X,Y>>>;
                                                                                                      ^
In file included from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/PreconditionerFactory.hpp:26,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/tests/test_preconditionerfactory.cpp:29:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp: In constructor ‘Dune::OwningTwoLevelPreconditioner<OperatorType, VectorType, transpose, Communication>::OwningTwoLevelPreconditioner(const OperatorType&, const pt&)’:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp:70:25: error: ‘PrecFactory’ has not been declared
         , finesmoother_(PrecFactory::create(linearoperator, prm.get_child("finesmoother")))
                         ^~~~~~~~~~~
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp: In constructor ‘Dune::OwningTwoLevelPreconditioner<OperatorType, VectorType, transpose, Communication>::OwningTwoLevelPreconditioner(const OperatorType&, const pt&, const Communication&)’:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp:95:25: error: ‘PrecFactory’ has not been declared
         , finesmoother_(PrecFactory::create(linearoperator, prm.get_child("finesmoother"), comm))
                         ^~~~~~~~~~~
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp: In member function ‘void Dune::OwningTwoLevelPreconditioner<OperatorType, VectorType, transpose, Communication>::updateImpl(const Comm*)’:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp:166:25: error: ‘PrecFactory’ has not been declared
         finesmoother_ = PrecFactory::create(linear_operator_, prm_.get_child("finesmoother"), *comm_);
                         ^~~~~~~~~~~
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp: In member function ‘void Dune::OwningTwoLevelPreconditioner<OperatorType, VectorType, transpose, Communication>::updateImpl(const Dune::Amg::SequentialInformation*)’:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp:173:25: error: ‘PrecFactory’ has not been declared
         finesmoother_ = PrecFactory::create(linear_operator_, prm_.get_child("finesmoother"));
                         ^~~~~~~~~~~
In file included from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/PreconditionerFactory.hpp:26,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/FlexibleSolver.hpp:24,
                 from /home/mblatt/src/dune/dune-2.7/opm-simulators/tests/test_flexiblesolver.cpp:29:
/home/mblatt/src/dune/dune-2.7/opm-simulators/opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp:45:7: error: conflicting declaration of template ‘template<class Operator, class Comm> struct Dune::PreconditionerFactory’
 class PreconditionerFactory;
       ^~~~~~~~~~~~~~~~~~~~~
...
```
I guess the easy fix would be to define the factories in the OPM namespace or is the reason to use the Dune namespace that I am failing to see?

**Comment by atgeirr** [comment]
> I guess the easy fix would be to define the factories in the OPM namespace or is the reason to use the Dune namespace that I am failing to see?

Using `namespace Opm` is the simple solution.

**Comment by blattms** [comment]
Closed via #2267
